Adding Bounding Boxes class to Base

GivAlz opened the pull request:
With this merge I'm adding a class, Bounding Boxes, to Base. It is meant to be simple and lightweight: the boundaries are represented by a couple of points and there's a member function, point_inside, which given a point returns true if there's a point inside the box, false otherwise.

This class is part of a bigger project I'm working at, as a student of @luca-heltai, which aims at providing new possibilities for fluid-structure interaction simulations. In particular they are meant to represent, approximately, the space occupied by the active cells of a mesh: more pull requests will follow for tools etc.

davydden commented:
please run indentation script with astyle 2.04

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+{
+public:
+  /**
+   * Standard constructor. Creates an object that corresponds to a null box,
```
~~null~ empty

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+public:
+  /**
+   * Standard constructor. Creates an object that corresponds to a null box,
+   * i.e. an degenerate box with both points being the origin.
```
A degenerate

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+   * which describe the box: one for the bottom and one for the top
+   * corner.
+   */
+  BoundingBox (const std::pair<Point<spacedim,Number>,Point<spacedim,Number>> &boundary_points); //Add two points constructor
```
remove the comment 
> //Add two points constructor

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+  /**
+   * Returns the boundary_points
+   */
+  std::pair<Point<spacedim,Number>,Point<spacedim,Number>> & get_boundary_points ();
```
should probably be `const`
```
const std::pair<Point<spacedim,Number>,Point<spacedim,Number>> & get_boundary_points () const;
```

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+ * Boundig boxes are represented by two vertices and, geometrically, are
+ * represented by the following:
+ * 1 d: a segment (represented by its vertices)
+ * 2 d: a rectangle (represented by two vertices V opposed to the center)
```
can user provide any pair or the one you describe here? 
i.e. bottom left + top right vs
bottom right + top left.

In dealii there is already a function https://www.dealii.org/developer/doxygen/deal.II/namespaceGridTools.html#a0374dd9eae73c3c2badaa479e8ba8d55
which defines it as bottom left + top right. So this can should be supported for consistency.

GivAlz replied:
I shall use the bottom left, top right convention then. This means I can also modify the is inside function.

davydden replied:
yes, and then it's also easy to check if the points satisfy bottom-left-to-right.

davydden commented on source/base/bounding_box.cc:
```diff
@@ -0,0 +1,50 @@
+}
+
+template <int spacedim, typename Number>
+bool is_degenerate () const
```
un-used, please remove

davydden commented on tests/base/bounding_box_1.cc:
```diff
@@ -0,0 +1,98 @@
+  std::pair<Point<spacedim>,Point<spacedim>> boundaries;
+
+  for(int i=0;i<spacedim;i++){
+      boundaries.first[i] += 0.2 -i*0.2;
```
i think `=` is easier to read here.

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+inline
+BoundingBox<spacedim, Number>::BoundingBox (const std::pair<Point<spacedim,Number>,Point<spacedim,Number>> &boundary_points)
+{
+    this->boundary_points = boundary_points;
```
i think you should check that points are correct.

GivAlz replied:
Thank you for the comments: they are very useful!!!
As for this one I wonder: this seems a right place to check if points generate a bounding box of dimension spacedim. Do you think this is the sort of test I should do?

The problem with an assert statement here is that dimension is guaranteed only when you create a bounding box giving a pair of points while with the default constructor you're guaranteed the bounding box is, in reality, a point...would this be ok?

Thank you very much for your help!

davydden replied:
>  Do you think this is the sort of test I should do?

yes, you need to check it here. Since you don't call this constructor from the default, it should be no problem.

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+inline
+BoundingBox<spacedim, Number>::BoundingBox (const std::pair<Point<spacedim,Number>,Point<spacedim,Number>> &boundary_points)
+{
+  this->boundary_points = boundary_points;
```
please check that provided points are correct.

davydden commented on tests/base/bounding_box_1.cc:
```diff
@@ -0,0 +1,103 @@
+
+  for (int i=0; i<spacedim; i++)
+    {
+      boundaries.first[i] += 0.2 -i*0.2;
```
=

davydden commented on tests/base/bounding_box_1.cc:
```diff
@@ -0,0 +1,103 @@
+  for (int i=0; i<spacedim; i++)
+    {
+      boundaries.first[i] += 0.2 -i*0.2;
+      boundaries.second[i] += 0.8 + i*0.8;
```
=

davydden commented on tests/base/bounding_box_1.cc:
```diff
@@ -0,0 +1,103 @@
+      Point<spacedim> test_pt;
+      c = (1+t)/5.0;
+      for (unsigned int i=0; i<spacedim; ++i)
+        {
```
you don't need brackets

davydden commented on tests/base/bounding_box_1.cc:
```diff
@@ -0,0 +1,103 @@
+      c = (1+t)*2.5;
+      if (t%2==0) c=-c; //Changing the sign sometimes..
+      for (unsigned int i=0; i<spacedim; ++i)
+        {
```
same

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+  /**
+   * Returns the boundary_points
+   */
+  std::pair<Point<spacedim,Number>,Point<spacedim,Number>> &get_boundary_points ();
```
either make this
```
const std::pair<Point<spacedim,Number>,Point<spacedim,Number>> &get_boundary_points () const;
```
or add the `const` version.

davydden replied:
on a second though, i think there should be no write access to points. If they are to be modified, then we need a separate method `set()` or something like that where we can make sure that the provided input makes sense.

GivAlz replied:
I agree: for what I need there's no need to modify them so I created only the const version.

GivAlz replied:
And thanks for the 200 comments, I didn't think the code would be so sloppy. I'll finish fixing and hopefully this part shall be ok

bangerth commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2016 by the deal.II authors
```
I think you want this to be 2017 only :-)

bangerth commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+ *
+ * Boundig boxes are represented by two vertices and, geometrically, are
+ * represented by the following:
+ * 1 d: a segment (represented by its vertices)
```
I think you want this to be a bullet point list. doxygen can do this for you if you start the line with a `-`.

bangerth commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+ * 2 d: a rectangle (represented by two vertices V opposed to the center)
+ * V--------.
+ * |        |
+ * .--------V
```
That's going to come out funny because doxygen doesn't respect line breaks. You'll want to wrap this in
```
@code
...ascii art...
@endcode
```

bangerth commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+ * .------. |
+ * |      | /
+ * |      |/
+ * .------V
```
same here

bangerth commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+inline
+BoundingBox<spacedim, Number>::BoundingBox ()
+{
+}
```
You can probably get away with just saying
```
  BoundingBox () = default;
```
in the class declaration. Then you don't have to write the definition of the constructor here.

bangerth commented on source/base/bounding_box.cc:
```diff
@@ -0,0 +1,43 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by Giovanni Alzetta
```
I'm afraid that if you contribute the code to the project, it needs to be "the deal.II authors" here as in all other files :-(

GivAlz replied:
Yes, sorry...that comes from a too stupid copy-paste from the other files (which is also revealing about my degree as "author" )

bangerth commented on source/base/bounding_box.cc:
```diff
@@ -0,0 +1,43 @@
+    {
+      //The point is outside the box if it is greater (or smaller) than both coordinates
+      if ( (this->boundary_points.first[i] < p[i] && this->boundary_points.second[i] < p[i])
+           || (this->boundary_points.first[i] > p[i] && this->boundary_points.second[i] > p[i]))
```
If you adopt the convention that the first point needs to be to the left of the right point, then you can avoid half of the comparisons here.

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,112 @@
+inline
+BoundingBox<spacedim, Number>::BoundingBox (const std::pair<Point<spacedim,Number>,Point<spacedim,Number>> &boundary_points)
+{
+  for (unsigned int i=0; i<spacedim; ++i)
```
i wonder what happens in `release` mode here as `Assert` is gone... @bangerth probably knows :smile: but one could do 
```
#ifdef DEBUG
   for ()

#endif
```
to be sure there are no warnings/etc.

bangerth replied:
No, we do this all the time without the `#ifdef`. In release mode, the compiler just sees
```
  for (unsigned int i=.....)
    ;
```
which it duly optimizes away completely because nothing in the loop header has any side effects. 

So this is fine.

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,112 @@
+BoundingBox<spacedim, Number>::BoundingBox (const std::pair<Point<spacedim,Number>,Point<spacedim,Number>> &boundary_points)
+{
+  for (unsigned int i=0; i<spacedim; ++i)
+    Assert (boundary_points.first[i] <= boundary_points.second[i],
```
<

bangerth replied:
But don't you want to allow degenerate boxes that are empty? In that case, `<=` is correct.

davydden replied:
agreed. Probably better to allow it so that one can have an empty box which is not at the origin.

GivAlz replied:
Degenerate boxes are empty, yes... but I have an empty constructor only because there was the comment in points.h about clang complaining if there's not one (otherwise I would eliminate it): when I create bounding boxes, for what I plan, I need them to have dimension spacedim: I guess < is the way to go.

davydden commented on source/base/bounding_box.cc:
```diff
@@ -0,0 +1,43 @@
+    {
+      //Bottom left-top right convention: the point is outside if it's smaller than the
+      //first or bigger than the second boundary point
+      if ( p[i] < this->boundary_points.first[i] || this->boundary_points.second[i] < p[i])
```
`<=` and `<=`

davydden replied:
also extend your first test to check that `0.,0.` is inside `0,0 -> 1,1` or alike.

bangerth replied:
I think it boils down to the question of whether you consider the bounding box to be mathematically open or closed. If you want to think of it as closed (for example because you want to allow bounding boxes that contain only a single point), then using `<` here is correct.

davydden replied:
correct. This should be mentioned in documentation. From practical purposes i think we want it to be closed (i.e. contain all its limit points).

GivAlz replied:
Yes, I need them to be closed, that's why I'm using <= (and in future a tolerance could be considered, effectively enlarging the box by 1e-10 or something...).

As for the test currently the first check it does is checking that boundary points are inside (i.e. 0.0. for 0,0-> 1,1); but I think it's a good idea to create a simple "unitary" box and check if all the vertices are included.

davydden commented on source/base/bounding_box.cc:
```diff
@@ -0,0 +1,43 @@
+  return this->boundary_points;
+}
+
+DEAL_II_NAMESPACE_CLOSE
```
not much need to close it just to open in the next line :smile:

davydden commented:
@GivAlz please add a changelog entry in `doc/news/changes/minor` section and squash the commits into a single one with meaningful description.

luca-heltai commented:
@GivAlz , thanks for putting up with my friends' comments, and don't be scared! The worst is yet to come... 

:)

luca-heltai commented on tests/base/bounding_box_1.cc:
```diff
@@ -0,0 +1,134 @@
+{
+  BoundingBox<spacedim> a;
+  deallog << "Empty constructor: \n";
+  deallog << a.get_boundary_points().first << "\n" << a.get_boundary_points().second << std::endl;
```
avoid using `\n` with deallog. Replace all `\n` with `std::endl`.

luca-heltai commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,104 @@
+ * Geometrically, are represent the following:
+ * 1 d: a segment (represented by its vertices)
+ * 2 d: a rectangle (represented by two vertices V opposed to the center)
+ * .--------V
```
Make sure you do 
~~~
@code
* .--------V
* |        |
* V--------.
@endcode
~~~

GivAlz replied:
Ok, I've slightly modified the description underlying the order must be respected (otherwise the point_inside function won't work).

Moreover there currently is an Assert which, when creating a BoundingBox from points, checks the order; do you think it'd be wiser to use an AssertThrow so that the check is done also at run time?

bangerth replied:
Nope, `Assert` is exactly the right choice here!

GivAlz commented:
Ok, I tried to fix everything; it went all well until I tried to squash commits making git extremely unhappy. Anyway everything but that should be ok...

GivAlz commented:
I had screw up with git log big-time (mixing my additions and some pulls from the dealii official repository) so I eventually did a hard reset. Sorry for that, but now there should be a single commit and I hope I caught all your useful comments.
Thank you for your help!
Giovanni

luca-heltai commented:
Since @GivAlz is my PhD student, I'll let someone else push the merge button. :)

masterleinad commented:
/run-tests

GivAlz commented:
Strange...the test that fails is "intergrid_constraints.debug" which I never touched nor I know what it is...

masterleinad commented:
@GivAlz I have seen this test failing with a timeout before. I am quite sure that this is unrelated.

masterleinad commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+ * description the owners of each portion of the mesh.
+ *
+ * Boundig boxes are represented by two vertices (bottom left and top right).
+ * Geometrically, are represent the following:
```
`Boundig`->`Bounding`
`Geometrically, a bounding box is `

masterleinad commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+{
+public:
+  /**
+   * Standard constructor. Creates an object that corresponds to a empty box,
```
an empty box

masterleinad commented on include/deal.II/base/bounding_box.h:
```diff
@@ -0,0 +1,108 @@
+public:
+  /**
+   * Standard constructor. Creates an object that corresponds to a empty box,
+   * i.e. an degenerate box with both points being the origin.
```
a degenerate box

masterleinad commented on doc/news/changes/minor/20170915GiovanniAlzetta:
```diff
@@ -0,0 +1,4 @@
+New: Added the base class BoundingBox. The BoundingBox is constructed from a std::pair of points in the space, ordered following the convention bottom-left, top-right.
```
of points in real space

masterleinad commented:
It would be great if you could the few small typos I found. Otherwise, this looks good to merge for me. I am looking forward to your next PRs!

GivAlz commented:
Ok, thank you for the grammar checks and all the help! I corrected and squashed...but it seems like my git history is not still perfect: if this is ok good...otherwise I can always checkout back in history and add the corrected files.

I'll add my next Pull Request soon...but before it needs to reach an "ok" level to my standards or we shall break the 100 comments mark ;)

bangerth commented:
Thanks for sticking with us through all of these comments!

I remember a (rather large) patch by @gassmoeller where I had 200+ comments. At that time, github still sent one email per comment. I'm pretty sure it took the poor man several days to get his inbox back into shape :-(

bangerth commented:
OK to merge once the tester is happy.
